File: go-aethermq/internal/raft/raft.go
package raft

import (
	"context"
	"sync"
	"time"

	"go.uber.org/zap"
)

type RaftNode struct {
	nodeID     string
	address    string
	state      RaftState
	leaderID   string
	term       uint64
	commitIdx  uint64
	lastApplied uint64

	log        []*LogEntry
	appliedCh  chan *LogEntry

	mu         sync.RWMutex
	logger     *zap.Logger
	running    bool
}

type RaftState int

const (
	Follower RaftState = iota
	Candidate
	Leader
)

type LogEntry struct {
	Index uint64
	Term  uint64
	Cmd   *Command
}

type Command struct {
	Op    OpType
	Key   string
	Value interface{}
}

type OpType int

const (
	CreateTopic OpType = iota
	DeleteTopic
	RegisterBroker
	UnregisterBroker
	UpdatePartition
	CommitOffset
)

func NewRaftNode(ctx context.Context, nodeID, address, dataDir, joinAddr string) (*RaftNode, error) {
	node := &RaftNode{
		nodeID:    nodeID,
		address:   address,
		state:     Follower,
		log:       make([]*LogEntry, 0),
		appliedCh: make(chan *LogEntry, 1000),
		logger:    zap.NewNop(),
	}

	go node.run(ctx)

	return node, nil
}

func (n *RaftNode) run(ctx context.Context) {
	n.running = true
	ticker := time.NewTicker(100 * time.Millisecond)
	defer ticker.Stop()

	for n.running {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			n.mu.Lock()
			switch n.state {
			case Follower:
				n.followerTick()
			case Candidate:
				n.candidateTick()
			case Leader:
				n.leaderTick()
			}
			n.mu.Unlock()
		}
	}
}

func (n *RaftNode) followerTick() {
}

func (n *RaftNode) candidateTick() {
	n.state = Leader
	n.leaderID = n.nodeID
	n.term++
	n.logger.Info("became leader", zap.Uint64("term", n.term))
}

func (n *RaftNode) leaderTick() {
}

func (n *RaftNode) Propose(ctx context.Context, cmd *Command) error {
	n.mu.Lock()
	defer n.mu.Unlock()

	if n.state != Leader {
		return ErrNotLeader
	}

	entry := &LogEntry{
		Index: uint64(len(n.log)),
		Term:  n.term,
		Cmd:   cmd,
	}
	n.log = append(n.log, entry)
	n.commitIdx = entry.Index
	n.lastApplied = entry.Index

	select {
	case n.appliedCh <- entry:
	default:
	}

	return nil
}

func (n *RaftNode) IsLeader() bool {
	n.mu.RLock()
	defer n.mu.RUnlock()
	return n.state == Leader
}

func (n *RaftNode) GetLeaderID() string {
	n.mu.RLock()
	defer n.mu.RUnlock()
	return n.leaderID
}

func (n *RaftNode) GetTerm() uint64 {
	n.mu.RLock()
	defer n.mu.RUnlock()
	return n.term
}

func (n *RaftNode) Close() {
	n.running = false
	close(n.appliedCh)
}

var ErrNotLeader = &RaftError{Message: "not the leader"}

type RaftError struct {
	Message string
}

func (e *RaftError) Error() string {
	return e.Message
}

File: rust-heliosmq/src/raft.rs
use std::sync::Arc;
use std::collections::HashMap;
use std::time::{Duration, Instant};

use parking_lot::{Mutex, RwLock};
use tokio::sync::mpsc;
use tracing::{info, warn, error};

use crate::config::ControllerConfig;

pub struct RaftNode {
    node_id: String,
    address: String,
    state: Mutex<RaftState>,
    leader_id: RwLock<Option<String>>,
    term: Mutex<u64>,
    commit_idx: Mutex<u64>,
    last_applied: Mutex<u64>,
    log: RwLock<Vec<LogEntry>>,
    applied_tx: mpsc::Sender<LogEntry>,
    running: Mutex<bool>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RaftState {
    Follower,
    Candidate,
    Leader,
}

#[derive(Debug, Clone)]
pub struct LogEntry {
    pub index: u64,
    pub term: u64,
    pub cmd: Command,
}

#[derive(Debug, Clone)]
pub struct Command {
    pub op: OpType,
    pub key: String,
    pub value: Vec<u8>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OpType {
    CreateTopic,
    DeleteTopic,
    RegisterBroker,
    UnregisterBroker,
    UpdatePartition,
    CommitOffset,
}

impl RaftNode {
    pub async fn new(config: &ControllerConfig) -> anyhow::Result<Self> {
        let (applied_tx, mut applied_rx) = mpsc::channel(1000);

        let node = Self {
            node_id: config.node_id.clone(),
            address: config.raft_addr.clone(),
            state: Mutex::new(RaftState::Follower),
            leader_id: RwLock::new(None),
            term: Mutex::new(0),
            commit_idx: Mutex::new(0),
            last_applied: Mutex::new(0),
            log: RwLock::new(Vec::new()),
            applied_tx,
            running: Mutex::new(true),
        };

        Ok(node)
    }

    pub fn is_leader(&self) -> bool {
        *self.state.lock() == RaftState::Leader
    }

    pub fn get_leader_id(&self) -> Option<String> {
        self.leader_id.read().clone()
    }

    pub fn get_term(&self) -> u64 {
        *self.term.lock()
    }

    pub async fn propose(&self, cmd: Command) -> anyhow::Result<()> {
        let mut state = self.state.lock();
        
        if *state != RaftState::Leader {
            return Err(anyhow::anyhow!("Not the leader"));
        }

        let term = *self.term.lock();
        let mut log = self.log.write();
        
        let entry = LogEntry {
            index: log.len() as u64,
            term,
            cmd,
        };

        log.push(entry.clone());
        
        *self.commit_idx.lock() = entry.index;
        *self.last_applied.lock() = entry.index;

        let _ = self.applied_tx.send(entry).await;

        Ok(())
    }

    pub fn close(&self) {
        *self.running.lock() = false;
    }
}
